### PR TITLE
Create camera

### PR DESCRIPTION
**Pull request by azharmalik3**
@mosic

### CODE REVIEW

File: web/controllers/camera_controller.ex [hunk 1]
```diff
@@ -56,6 +59,37 @@ defmodule EvercamMedia.CameraController do
...
   end
 
+  def create(conn, params) do
+    try do
+      caller = conn.assigns[:current_user]
+      if is_nil(caller), do: [401, %{message: "Unauthorized"}]
+      if is_nil(params["id"]) do
+        params = Map.merge(%{"id" => auto_generate_camera_id(params["name"]), "owner_id" => caller.id}, params)
+      else
+        String.downcase(params["id"])
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: This line doesn't have any effect.

File: web/controllers/camera_controller.ex [hunk 2]
```diff
@@ -56,6 +59,37 @@ defmodule EvercamMedia.CameraController do
     end
   end
 
+  def create(conn, params) do
+    try do
+      caller = conn.assigns[:current_user]
+      if is_nil(caller), do: [401, %{message: "Unauthorized"}]
+      if is_nil(params["id"]) do
+        params = Map.merge(%{"id" => auto_generate_camera_id(params["name"]), "owner_id" => caller.id}, params)
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: `owner_id` is being added below this in all cases, no need to do it here also.

File: web/controllers/camera_controller.ex [hunk 3]
```diff
@@ -56,6 +59,37 @@ defmodule EvercamMedia.CameraController do
...
+      end
+      params = Map.merge(%{"owner_id" => caller.id}, params)
+
+      case validate_params(params) do
+        {:invalid, message} ->
+          add_render(conn, 400, %{message: message})
+        {:duplicate, message} ->
+          add_render(conn, 409, %{message: message})
+        {:ok} ->
+          camera =
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: No need for a pipeline here, just use `camera = camera_create(params)`.

File: web/controllers/camera_controller.ex [hunk 4]
```diff
@@ -56,6 +59,37 @@ defmodule EvercamMedia.CameraController do
...
+            params
+            |> camera_create
+          conn
+            |> put_status(200)
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Correct status when creating a resource is 201.

And it can probably be omitted completely, I'd expect Phoenix to put the correct status because this is a POST method.

File: web/controllers/camera_controller.ex [hunk 5]
```diff
@@ -56,6 +59,37 @@ defmodule EvercamMedia.CameraController do
...
+            |> render("show.json", %{camera: camera, user: caller})
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: This should line up with `conn`, it doesn't need indentation.

File: web/controllers/camera_controller.ex [hunk 6]
```diff
@@ -56,6 +59,37 @@ defmodule EvercamMedia.CameraController do
...
+      end
+    rescue
+      error ->
+        Logger.error "[camera create] [error] [#{error}]"
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Trying to log the error here would fail with an error (ironic?). Try it and see.

File: web/controllers/camera_controller.ex [hunk 7]
```diff
@@ -56,6 +59,37 @@ defmodule EvercamMedia.CameraController do
...
+        add_render(conn, 500, %{message: "Sorry, we dropped the ball."})
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: There is no need for `add_render` function, just use `ErrorView`.

File: web/controllers/camera_controller.ex [hunk 8]
```diff
@@ -112,4 +146,110 @@ defmodule EvercamMedia.CameraController do
...
   end
+
+  defp auto_generate_camera_id(camera_name) do
+    camera_name =
+      camera_name
+      |> String.replace(" ", "")
+      |> String.replace("-", "")
+      |> String.downcase
+      |> String.slice(0..4)
+      "#{camera_name}-#{Enum.take_random(?a..?z, 5)}"
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Unindent this, it's not a part of the pipeline.

File: web/controllers/camera_controller.ex [hunk 9]
```diff
@@ -112,4 +146,110 @@ defmodule EvercamMedia.CameraController do
         Logger.info "Skipping camera worker update as the host is invalid"
     end
   end
+
+  defp auto_generate_camera_id(camera_name) do
+    camera_name =
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: It would be slightly more clear if this variable was called `camera_id` instead of `camera_name`.

File: web/controllers/camera_controller.ex [hunk 10]
```diff
@@ -112,4 +146,110 @@ defmodule EvercamMedia.CameraController do
...
+        {:invalid, "External url is invalid"}
+      params["model"] && !params["vendor"] ->
+        {:invalid, "If you provide model you must also provide vendor"}
+      params["mac_address"] && !Regex.match?(~r/^([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})$/, params["mac_address"]) ->
+        {:invalid, "Mac address is invalid"}
+      params["location_lng"] && is_nil(params["location_lat"]) ->
+        {:invalid, "Must provide both location coordinates"}
+      params["location_lat"] && is_nil(params["location_lng"]) ->
+        {:invalid, "Must provide both location coordinates"}
+      params["timezone"] && false == Tzdata.zone_exists?(params["timezone"]) ->
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: `false == something` is confusing, either do `!something` or `something == false`.

File: web/controllers/camera_controller.ex [hunk 11]
```diff
@@ -112,4 +146,110 @@ defmodule EvercamMedia.CameraController do
...
+    if params["location_lng"] && params["location_lat"] do
+      camera_params = Map.put(camera_params, :location, %Geo.Point{coordinates: {params["location_lng"], params["location_lat"]}})
+    end
+
+    if model, do: camera_params = Map.put(camera_params, :model_id, model.id)
+
+    if params["mac_address"], do: camera_params = Map.put(camera_params, :mac_address, params["mac_address"])
+    if params["timezone"], do: camera_params = Map.put(camera_params, :timezone, params["timezone"])
+
+    changeset = Camera.changeset(%Camera{}, camera_params)
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: This can be modified to be a nice long pipeline like this:

``` elixir
%Camera{}
|> Camera.changeset(camera_params)
|> Repo.insert!
|> Repo.preload(:owner)
|> Repo.preload(:vendor_model)
|> Repo.preload([vendor_model: :vendor])
```

File: web/controllers/camera_controller.ex [hunk 12]
```diff
@@ -112,4 +146,110 @@ defmodule EvercamMedia.CameraController do
...
+    Repo.insert!(changeset)
+    |> Repo.preload([:owner])
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: There is no need for `[]` around `:owner`.

File: web/controllers/camera_controller.ex [hunk 13]
```diff
@@ -112,4 +146,110 @@ defmodule EvercamMedia.CameraController do
...
+    |> Repo.preload(:vendor_model)
+    |> Repo.preload([vendor_model: :vendor])
+  end
+
+  ######################
+  ## Render functions ##
+  ######################
+
+  defp add_render(conn, code, response) do
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Remove this as mentioned above.

File: web/controllers/camera_controller.ex [hunk 14]
```diff
@@ -112,4 +146,110 @@ defmodule EvercamMedia.CameraController do
...
+        {:invalid, "Timezone does not exist or is invalid"}
+      is_nil(Camera.by_exid(params["id"])) == false ->
+        {:duplicate, "Camera id already exists."}
+      true ->
+        {:ok}
+    end
+  end
+
+  defp camera_create(params) do
+    vendor = model = nil
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Don't do multiple assignments. It's a bit shorter but is also needlessly confusing.

File: web/controllers/camera_controller.ex [hunk 15]
```diff
@@ -112,4 +146,110 @@ defmodule EvercamMedia.CameraController do
...
+    if params["vendor"] do
+      vendor = Vendor.by_exid(String.downcase(params["vendor"]))
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: This is nested two times, looks like a good opportunity for a pipeline.

File: web/controllers/camera_controller.ex (hunk 14 above)
Comment (mosic) [inline web/controllers/camera_controller.ex]: Also don't assign nil to a variable.

File: web/controllers/camera_controller.ex [hunk 16]
```diff
@@ -112,4 +146,110 @@ defmodule EvercamMedia.CameraController do
...
+    end
+    cond do
+      params["model"] ->
+        model = VendorModel.by_exid(String.downcase(params["model"]))
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: This won't work in future versions of Elixir, read more about it here: http://tuvistavie.com/2016/elixir-1-3/#deprecation-of-imperative-assignment

File: web/controllers/camera_controller.ex [hunk 17]
```diff
@@ -112,4 +146,110 @@ defmodule EvercamMedia.CameraController do
...
+    cond do
+      Regex.match?(~r/^[a-z0-9\-_]+$/, params["id"]) == false || String.length(params["id"]) < 3 ->
+        {:invalid, "Camera ID can only contain lower case letters, numbers, hyphens and underscore. Minimum length is 4."}
+      String.length(params["name"]) > 25 ->
+        {:invalid, "Camera Name is too long. Maximum 24 characters."}
+      Regex.match?(@ip_address_regex, params["external_host"]) == false or Regex.match?(@hostname_regex, params["external_host"]) == false ->
+        {:invalid, "External url is invalid"}
+      params["model"] && !params["vendor"] ->
+        {:invalid, "If you provide model you must also provide vendor"}
+      params["mac_address"] && !Regex.match?(~r/^([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})$/, params["mac_address"]) ->
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Make `mac_address` a module variable, like you did for `ip_address` and `hostname`.

File: web/controllers/camera_controller.ex [hunk 18]
```diff
@@ -56,6 +59,37 @@ defmodule EvercamMedia.CameraController do
     end
   end
 
+  def create(conn, params) do
+    try do
+      caller = conn.assigns[:current_user]
+      if is_nil(caller), do: [401, %{message: "Unauthorized"}]
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: This doesn't do anything.

File: web/controllers/camera_controller.ex [hunk 19]
```diff
@@ -112,4 +146,110 @@ defmodule EvercamMedia.CameraController do
...
+        model = VendorModel.get_vendor_default_model(vendor)
+      true ->
+        model = VendorModel.by_exid("other_default")
+    end
+
+    camera_params = %{
+      exid: "#{params["id"]}",
+      name: params["name"],
+      owner_id: params["owner_id"],
+      is_public: params["is_public"],
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: This is expecting that the `is_public` parameter exists, but you're not validating that beforehand. One way to deal with this is to default to `false` and replace the value if provided.

File: web/controllers/camera_controller.ex (hunk 18 above)
Comment (azharmalik3) [inline web/controllers/camera_controller.ex]: @mosic if i am not using this endpoint gave response `500 internal server error`. Because you considered `valid` if no `api_id` and `api_key`. https://github.com/evercam/evercam-media/blob/master/web/auth.ex#L2

That's why i am using this line.

File: web/controllers/camera_controller.ex [hunk 20]
```diff
@@ -112,4 +153,106 @@ defmodule EvercamMedia.CameraController do
...
+    end
+  end
+
+  defp camera_create(params) do
+    vendor =  
+      if params["vendor"] do
+        params["vendor"]
+        |> String.downcase
+        |> Vendor.by_exid
+      end
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: What if `params["vendor"]` isn't provided? You need to define all cases.

File: web/controllers/camera_controller.ex [hunk 21]
```diff
@@ -112,4 +153,106 @@ defmodule EvercamMedia.CameraController do
...
+    model = 
+      cond do
+        params["model"] ->
+          params["model"]
+          |> String.downcase
+          |> VendorModel.by_exid
+        !is_nil(vendor) ->
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Why not just `vendor`?

File: web/controllers/camera_controller.ex [hunk 22]
```diff
@@ -112,4 +153,106 @@ defmodule EvercamMedia.CameraController do
...
+      |> String.downcase
+      |> String.slice(0..4)
+    "#{camera_id}-#{Enum.take_random(?a..?z, 5)}"
+  end
+
+  defp validate_params(params) do
+    cond do
+      !Regex.match?(~r/^[a-z0-9\-_]+$/, params["id"]) || String.length(params["id"]) < 3 ->
+        {:invalid, "Camera ID can only contain lower case letters, numbers, hyphens and underscore. Minimum length is 4."}
+      String.length(params["name"]) > 25 ->
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: If the length is maximum 24 characters, this should be `> 24`.

File: web/controllers/camera_controller.ex [hunk 23]
```diff
@@ -56,6 +60,43 @@ defmodule EvercamMedia.CameraController do
...
+          |> render(ErrorView, "error.json", %{message: message})
+        {:ok} ->
+          camera = camera_create(params)
+          conn
+          |> render("show.json", %{camera: camera, user: caller})
+      end
+    rescue
+      error ->
+        Logger.error "[camera create] [error] [#{inspect(error)}]"
+      conn
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: This is inside the `error ->` block, so it should line up with `Logger` line.

File: web/controllers/camera_controller.ex (hunk 18 above)
Comment (mosic) [inline web/controllers/camera_controller.ex]: @azharmalik3 My point still stands, the line doesn't have any effect. I'm guessing you're expecting it to cause an early return, but it doesn't work that way.

**Comment by mosic** [comment]
@azharmalik3 There are still fundamental problems with this PR. I suggest writing tests for it - one with correct parameters and a couple for different ways the request can fail.

That way you should be able to see where the issues are.

File: web/controllers/camera_controller.ex (hunk 18 above)
Comment (azharmalik3) [inline web/controllers/camera_controller.ex]: @mosic It will raise an error when `current_user` is nill.

File: web/controllers/camera_controller.ex (hunk 21 above)
Comment (azharmalik3) [inline web/controllers/camera_controller.ex]: @mosic It will raise an error in case of `params["vendor"]` not provided.

File: web/controllers/camera_controller.ex (hunk 20 above)
Comment (azharmalik3) [inline web/controllers/camera_controller.ex]: @mosic Below all code will work fine no matter vendor provided or not.

Comment (mosic) [inline web/controllers/camera_controller.ex]: @azharmalik3 That's because it's relying on the fact that vendor will be given `nil`. It's always better to explicitly specify possible values. You should add an `else` clause.

File: web/controllers/camera_controller.ex (hunk 21 above)
Comment (mosic) [inline web/controllers/camera_controller.ex]: @azharmalik3 That wouldn't happen if the vendor was explicitly defined as I mentioned above.

File: web/controllers/camera_controller.ex (hunk 18 above)
Comment (mosic) [inline web/controllers/camera_controller.ex]: @azharmalik3 I don't understand why are you so insistent on this. It won't raise an error, try running that line of code and see for yourself.

The only thing it does is returns a list with the status and message, which isn't assigned and won't be used anywhere.

File: web/models/camera.ex [hunk 24]
```diff
@@ -142,16 +142,16 @@ defmodule Camera do
...
+    case external_url(camera, network) != "" && res_url(camera, type) != "" do
+      true -> external_url(camera, network) <> res_url(camera, type)
       false -> ""
     end
   end
 
-  def external_url(camera, protocol \\ "http") do
+  def external_url(camera, network, protocol \\ "http") do
     host = host(camera) |> to_string
-    port = port(camera, "external", protocol) |> to_string
```
Comment (mosic) [inline web/models/camera.ex]: Revert this change, it's completely misleading. `external_url` function shouldn't be able to show internal url, so there's no reason to add the network parameter.

File: web/controllers/camera_controller.ex [hunk 25]
```diff
@@ -214,30 +223,30 @@ defmodule EvercamMedia.CameraController do
...
-
-    if params["jpg_url"], do: camera_params = put_in(camera_params, [:config, :snapshots, :jpg_url], params["jpg_url"])
-    if params["mjpg_url"], do: camera_params = put_in(camera_params, [:config, :snapshots, :mjpg_url], params["mjpg_url"])
-    if params["h264_url"], do: camera_params = put_in(camera_params, [:config, :snapshots, :h264_url], params["h264_url"])
-    if params["audio_url"], do: camera_params = put_in(camera_params, [:config, :snapshots, :audio_url], params["audio_url"])
-    if params["mpeg_url"], do: camera_params = put_in(camera_params, [:config, :snapshots, :mpeg_url], params["mpeg_url"])
+    if params["internal_host"], do: camera_params = put_in(camera_params, [:config, "internal_host"], params["internal_host"])
+    if params["internal_http_port"], do: camera_params = put_in(camera_params, [:config, "internal_http_port"], params["internal_http_port"])
+    if params["internal_rtsp_port"], do: camera_params = put_in(camera_params, [:config, "internal_rtsp_port"], params["internal_rtsp_port"])
+
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Remove whitespace.

File: web/models/camera.ex (hunk 24 above)
Comment (mosic) [inline web/models/camera.ex]: This needs to be changed everywhere it was used of course.

File: web/controllers/camera_controller.ex [hunk 26]
```diff
@@ -214,30 +223,30 @@ defmodule EvercamMedia.CameraController do
       owner_id: params["owner_id"],
       is_public: params["is_public"],
       config: %{
-        external_host: params["external_host"],
-        snapshots: %{}
+        "external_host" => params["external_host"],
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: What's the reason for this change?

**Comment by azharmalik3** [comment]
@mosic I just reverted external_url functions and added separate function for internal URLs. Now, I think it's ready to merge.

**Comment by mosic** [comment]
@azharmalik3 This branch seems to have diverged from master. Can you rebase it on top of current master? Also squash changes to one commit.

**Comment by azharmalik3** [comment]
@mosic I have added tests for camera create endpoint.

**Comment by azharmalik3** [comment]
@mosic I have squashed all commits into one commit.

**Comment by marcoherbst** [comment]
@azharmalik3 @mosic ... 2 days ago ? .... where is this being delayed?

**Comment by azharmalik3** [comment]
@marcoherbst just waiting for @mosic thumbs up.

**Comment by azharmalik3** [comment]
@mosic I have removed all trailing spaces and I think it's ready to merge.

File: web/controllers/camera_controller.ex [hunk 27]
```diff
@@ -112,6 +149,59 @@ defmodule EvercamMedia.CameraController do
...
+      name: params["name"],
+      owner_id: params["owner_id"],
+      is_public: params["is_public"],
+      config: %{
+        "external_host" => params["external_host"],
+        "snapshots" => %{}
+      }
+    }
+
+    if params["external_http_port"], do: camera_params = put_in(camera_params, [:config, "external_http_port"], params["external_http_port"])
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Same comment as this one: https://github.com/evercam/evercam-media/pull/243#discussion_r66449996

```
Unsafe variable found at:
  web/controllers/camera_controller.ex:194

warning: the variable "camera_params" is unsafe as it has been set inside a case/cond/receive/if/&&/||. Please explicitly return the variable value instead. For example:

    case int do
      1 -> atom = :one
      2 -> atom = :two
    end

should be written as

    atom =
      case int do
        1 -> :one
        2 -> :two
      end
```

This applies to all lines with `params` below.

File: web/controllers/camera_controller.ex [hunk 28]
```diff
@@ -69,6 +70,42 @@ defmodule EvercamMedia.CameraController do
     end
   end
 
+  def create(conn, params) do
+    try do
+      caller = conn.assigns[:current_user]
+      if caller do
+        unless params["is_public"], do: params = Map.merge(%{"is_public" => "false"}, params)
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Same as https://github.com/evercam/evercam-media/pull/212/files#r66459693

**Comment by azharmalik3** [comment]
Will be recreate.

**Comment by mosic** [comment]
@azharmalik3 Good idea to close this. I think you should pause on the `POST /v1/cameras` until `PUT /v1/cameras` is approved, so you don't have to make the changes in two places.
